### PR TITLE
Use num_qubits throughout Terra

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

In Qiskit the number of qubits of objects is referred to in different fashions, `n_qubits`/`num_qubits`/`numberofqubits`/etc. This PR implements the attribute `num_qubits` for all objects and deprecates the previous attribute.

#### Why `num_qubits` over all the others?

1. Because it is more descriptive than `n_qubits` or `n_qargs` and is less cumbersome than `numberofqubits`.
1. It's the fewest changes, currently the occurances of the different arguments in all the Qiskit repositories are (according to a `grep` search)
```
n_qubits | num_qubits | numberofqubits | n_qarg | num_qarg
---------+------------+----------------+--------+----------
145      | 371        | 6              | 2      | 4
```

#### JSON Schemas

I updated the JSON schemas as well, but in a discussion yesterday I realized they probably don't have to/cannot change due to the IQX API. Theses changes will be reverted.

#### Todos

- [x] revert changes on JSON schemas
- [x] implement tests checking the deprecated arguments are still supported 
**Edit:** resolved since the only deprecation is `n_qubits` in `QuantumCircuit` and `numberofqubits` in `Pauli`. The backward compatibility would've been more troublesome if the JSON schemas changed.

### CODE REVIEW

**Comment by taalexander** [comment]
I personally prefer `n_qubits` as this is the standard from the schemas.

**Comment by Cryoris** [comment]
@taalexander I see that it would be nice to preserve the `n_qubits` from the schemas and I'd prefer it to be the same all over in Qiskit. However the naming convention `num_<something>` is by far more common:
- Terra: `num_ctrl_qubits`, `num_ancillas`, `num_clbits`, `num_circuits`, etc.
- Aqua: `num_parameters`, `num_atoms`, `num_orbitals`, `num_alpha`, `num_beta`
- Aer: actually uses `num_qubits` in the C++ codes a lot, also `num_states`
- Ignis: `num_runs`, `num_of_states`, `num_of_gates`

It would be strange for the number of qubits to be the only thing called `n_qubits`.

**Comment by nonhermitian** [comment]
This makes sense. However backends return n_qubits per the spec. So either all the configs need to be changed, or your going to have to make a conversion step. Both are relatively unsatisfactory.

**Comment by chriseclectic** [comment]
I think choosing either `num_*` or  `n_*` and being consistent everywhere would be good.

Say we chose `num_*`  then we should change things like `num_of_states` to just `num_states`, and `n_qargs` to num_qargs`. Similarly if we chose `n_*` and rename all those other `num_thing`s to `n_thing` instead.

`n_qargs / num_qargs` still has a place so those don't necessarily need to be changed to `n_qubits` (not everything is qubits, but maybe something that has qargs and also qubits should have both and they return the same value).

**Comment by Cryoris** [comment]
> This makes sense. However backends return n_qubits per the spec. So either all the configs need to be changed, or your going to have to make a conversion step. Both are relatively unsatisfactory.

It seems we get away with almost no conversions between `num_qubits` and `n_qubits`, there are only a handful places where this is needed. 

E.g. one is needed in the circuit assembler:
```python
header = QobjExperimentHeader(..., n_qubits=num_qubits,...)
```
and in a few places we have
```python
num_qubits = config.n_qubits
num_qubits = config['n_qubits']
```

If you have a look at the changed files these conversions happen in very few places only.

File: qiskit/circuit/library/boolean_logic.py [hunk 1]
```diff
@@ -21,41 +21,46 @@
...
     """An n_qubit circuit that permutes qubits."""
 
+    @deprecate_arguments({'n_qubits': 'num_qubits'})
     def __init__(self,
-                 n_qubits: int,
+                 num_qubits: int,
                  pattern: Optional[List[int]] = None,
-                 seed: Optional[int] = None) -> QuantumCircuit:
+                 seed: Optional[int] = None,
+                 *, n_qubits: Optional[int] = None  # pylint:disable=unused-argument
```
Comment (taalexander) [inline qiskit/circuit/library/boolean_logic.py]: Could you please help me understand the pattern I see here and its purpose? I see it repeated a number of times when used with `deprecate_arguments` but not everywhere.

Comment (Cryoris) [reply qiskit/circuit/library/boolean_logic.py]: It's an additional safeguard as far as I understand, to prevent accidental setting of `num_qubits` and `n_qubits` as unnamed args. 
E.g. `Permutation(num_qubits, pattern, seed, num_qubits)` would raise an error since `deprecate_arguments` maps `n_qubits` (the last argument) to `num_qubits` which has already been provided (first argument) and raises. This might occur in a test case when the number of arguments is inferred via the signature. Pretty much an edge-case though and we can probably safely remove it :slightly_smiling_face:

File: qiskit/circuit/quantumcircuit.py [hunk 2]
```diff
@@ -960,22 +960,36 @@ def width(self):
...
         return qubits
 
     @property
-    def n_clbits(self):
-        """
-        Return number of classical bits.
-        """
+    def n_qubits(self):
+        """Deprecated, use ``num_qubits`` instead. Return number of qubits."""
+        warnings.warn('The QuantumCircuit.n_qubits method is deprecated as of 0.14.0, and '
```
Comment (mtreinish) [inline qiskit/circuit/quantumcircuit.py]: The next release is currently 0.13.0? Are we going to defer this until after 0.13.0, if so then we should label this on hold. Otherwise lets change this to 0.13.0

File: qiskit/extensions/standard/ms.py [hunk 3]
```diff
@@ -28,14 +28,16 @@
...
 from qiskit.circuit import QuantumRegister
+from qiskit.util import deprecate_arguments
 
 
 class MSGate(Gate):
     """Global Molmer-Sorensen gate."""
 
-    def __init__(self, n_qubits, theta):
+    @deprecate_arguments({'n_qubits': 'num_qubits'})
+    def __init__(self, num_qubits, theta, n_qubits=None):  # pylint:disable=unused-argument
```
Comment (mtreinish) [inline qiskit/extensions/standard/ms.py]: do we need the `*` here to enforce n_qubits can only be passed in by kwarg?

File: qiskit/transpiler/passes/layout/dense_layout.py [hunk 4]
```diff
@@ -115,18 +116,20 @@ def run(self, dag):
                 map_iter += 1
         self.property_set['layout'] = layout
 
-    def _best_subset(self, n_qubits):
+    @deprecate_arguments({'n_qubits': 'num_qubits'})
```
Comment (mtreinish) [inline qiskit/transpiler/passes/layout/dense_layout.py]: You don't need to bother with this here, it's a private method in a pass. If someone is using this outside of dense_layout they can not expect this to have a stable interface, especially if they're passing in the sole positional argument by name. Lets just swap it out and remove all the deprecation.

File: releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+deprecations:
+  - |
+    The ``n_qubits`` and ``numberofqubits`` keywords are deprecated throughout 
+    Terra and replaced by ``num_qubits``.
```
Comment (mtreinish) [inline releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml]: Honestly, I know it's a real pain but for this kind of thing it's good to list out exactly which user facing api has the deprecated arg and what they should migrate it to.

File: qiskit/circuit/library/boolean_logic.py [hunk 6]
```diff
@@ -21,41 +21,46 @@
 import numpy as np
 from qiskit.circuit import QuantumRegister, QuantumCircuit
 from qiskit.circuit.exceptions import CircuitError
+from qiskit.util import deprecate_arguments
 
 
 class Permutation(QuantumCircuit):
     """An n_qubit circuit that permutes qubits."""
 
+    @deprecate_arguments({'n_qubits': 'num_qubits'})
```
Comment (mtreinish) [inline qiskit/circuit/library/boolean_logic.py]: I believe this hasn't been released yet so you don't have to deprecate the arg we can just change it up until the release

File: qiskit/circuit/library/boolean_logic.py [hunk 7]
```diff
@@ -69,33 +74,37 @@ class XOR(QuantumCircuit):
     This circuit can also represent addition by ``amount`` over the finite field GF(2).
     """
 
+    @deprecate_arguments({'n_qubits': 'num_qubits'})
```
Comment (mtreinish) [inline qiskit/circuit/library/boolean_logic.py]: Same here, no need to deprecate as this hasn't been released

File: qiskit/circuit/library/boolean_logic.py [hunk 8]
```diff
@@ -105,20 +114,24 @@ def __init__(self,
 class InnerProduct(QuantumCircuit):
     """An n_qubit circuit that computes the inner product of two registers."""
 
-    def __init__(self, n_qubits: int) -> QuantumCircuit:
+    @deprecate_arguments({'n_qubits': 'num_qubits'})
```
Comment (mtreinish) [inline qiskit/circuit/library/boolean_logic.py]: and here too

File: qiskit/circuit/quantumcircuit.py (hunk 2 above)
Comment (Cryoris) [reply qiskit/circuit/quantumcircuit.py]: Yes, you're right

File: qiskit/extensions/standard/ms.py (hunk 3 above)
Comment (Cryoris) [reply qiskit/extensions/standard/ms.py]: I didn't put it there because then the `MSGate` required special treatment in a test where the number of arguments was inferred via the signature of `__init__`. I added the special treatment and `n_qubits` is now keyword only.

File: releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml (hunk 5 above)
Comment (Cryoris) [reply releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml]: I added a more detailed table :+1:

File: releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+deprecations:
+  - |
+    The ``n_qubits`` and ``numberofqubits`` keywords are deprecated throughout
+    Terra and replaced by ``num_qubits``. The objects affected by this change are listed below.
+
+    class                  | old method      | new method
+    -----------------------+-----------------+--------------------------------
+    QuantumCircuit         | n_qubits        | num_qubits
+    Pauli                  | numberofqubits  | num_qubits
+    BackendConfiguration   | nothing removed | num_qubits (additionally added)
```
Comment (mtreinish) [inline releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml]: This would be a new feature not a deprecation if we wanted to advertise. I'm not sure it's really a change for something that we need to document to users as a change in the next release.

**Comment by ajavadia** [comment]
This looks good so let's get it in the next release and start the deprecation period.

File: releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml (hunk 9 above)
Comment (Cryoris) [reply releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml]: Updated 👍

File: qiskit/quantum_info/operators/pauli.py [hunk 10]
```diff
@@ -256,10 +258,19 @@ def sgn_prod(p1, p2):
...
     @property
-    def numberofqubits(self):
+    def num_qubits(self):
         """Number of qubits."""
         return len(self)
 
+    @property
+    def numberofqubits(self):
+        """Deprecated, use ``num_qubits`` instead. Number of qubits."""
+        warnings.warn('The Pauli.numberofqubits method is deprecated as of 0.14.0, and '
```
Comment (mtreinish) [inline qiskit/quantum_info/operators/pauli.py]: We're still targetting this for 0.13.0:

```suggestion
        warnings.warn('The Pauli.numberofqubits method is deprecated as of 0.13.0, and '
```

File: releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml (hunk 9 above)
Comment (mtreinish) [reply releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml]: You actually were closer before, the other lines should be a deprecation. I specifically meant that the `BackendConfiguration` line in this table is not a deprecation. You're adding a new attribute/property to the class. The others are definitely deprecations

Comment (Cryoris) [reply releasenotes/notes/num-qubits-streamlining-01a9f58684ca4609.yaml]: Ah, now I understood. Yes that makes more sense 🙂

File: qiskit/quantum_info/operators/pauli.py (hunk 10 above)
Comment (Cryoris) [reply qiskit/quantum_info/operators/pauli.py]: Thanks for noticing 👍
